### PR TITLE
Adding Custom Pre- and Post- Log Hooks

### PR DESCRIPTION
**Pull request by aqeelat**
Rebase of #275

### CODE REVIEW

File: auditlog/diff.py [hunk 1]
```diff
@@ -98,7 +99,9 @@ def mask_str(value: str) -> str:
     return "*" * mask_limit + value[mask_limit:]
 
 
-def model_instance_diff(old, new, fields_to_check=None):
+def model_instance_diff(
+    old: Union[Model, None], new: Union[Model, None], fields_to_check=None
```
Comment (hramezani) [inline auditlog/diff.py:103]: Please use `Optional` instead of `Union` here

File: auditlog_tests/tests.py [hunk 2]
```diff
@@ -1862,6 +1865,159 @@ def test_access_log(self):
...
+        pre_log.connect(pre_log_receiver)
+        post_log.connect(post_log_receiver)
+
+        self.obj.text = "Changed Text"
+        self.obj.save()
+
+        self.assertSignals(LogEntry.Action.UPDATE)
+
+    @patch("auditlog.receivers.LogEntry.objects")
+    def test_signals_errors(self, log_entry_objects_mock):
```
Comment (hramezani) [inline auditlog_tests/tests.py:1995]: it would be great to test also the `post_log` signal here

File: docs/source/internals.rst [hunk 3]
```diff
@@ -25,6 +25,16 @@ Signal receivers
 .. automodule:: auditlog.receivers
     :members:
 
+Custom Signals
```
Comment (hramezani) [inline docs/source/internals.rst:28]: It would be great to have a `versionadded` tag here

File: auditlog_tests/tests.py (hunk 2 above)
Comment (aqeelat) [reply auditlog_tests/tests.py:1995]: They're all tested already.

Comment (hramezani) [reply auditlog_tests/tests.py:2044]: I mean the error in post_log. Did I miss it?

**Comment by aqeelat** [comment]
I’m not sure that I follow. post_log doesn’t raise errors. The error could be raised in getting the diff or in creating the log entry, not from the signals.Sent from my iPhoneOn Dec 24, 2022, at 3:06 PM, Hasan Ramezani ***@***.***> wrote:﻿
@hramezani commented on this pull request.



In auditlog_tests/tests.py:
> +            self.my_post_log_data["is_called"] = True
+            self.my_post_log_data["my_sender"] = sender
+            self.my_post_log_data["my_instance"] = instance
+            self.my_post_log_data["my_action"] = action
+            self.my_post_log_data["my_error"] = error
+
+        pre_log.connect(pre_log_receiver)
+        post_log.connect(post_log_receiver)
+
+        self.obj.text = "Changed Text"
+        self.obj.save()
+
+        self.assertSignals(LogEntry.Action.UPDATE)
+
+    @patch("auditlog.receivers.LogEntry.objects")
+    def test_signals_errors(self, log_entry_objects_mock):

I mean the error in post_log. Did I miss it?

**Comment by hramezani** [comment]
You will send the error to signal handler, I mean testing the error that is sent to the signal handler.
You did it in previous tests

File: auditlog_tests/tests.py (hunk 2 above)
Comment (aqeelat) [reply auditlog_tests/tests.py:2044]: Oh, I see what you mean. Tbh, initially, I just copied the tests as-is from the other PR. I just went over it now and updated a few things and added the test you suggested.

**Review by hramezani — APPROVED** [review]

**Comment by hramezani** [comment]
Thanks @aqeelat

**Pull request merged**